### MINIST is called Hello world of Deep Learning

### Wonderful book on deep learning: https://livebook.manning.com/book/deep-learning-with-python/chapter-1/143

I am finding this book, deep-learning with python, to be very fascinating. I think I will finish this book very soon (estimated by how eagerly I am reading it :))

This book was written by the Keras creator and Google AI researcher Francois Chollet.

## Chapter-2 : Mathematical Building Blocks of Neural Networks

### Keywords: Tensors, Tensor Operations, Differentiation, Gradient Descent


Now, the book drive into the example of digit classifier from Kaggle. The MNIST dataset for digit classification (0 to 9) consists of 60,000 training images and above 10, 000 test images.



### Loading the MNIST dataset in Keras

In [5]:
#!pip install keras


11493376/11490434 [==============================] - 14s 1us/step


In [6]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [7]:
train_images.shape

(60000, 28, 28)

In [8]:
len (train_labels)

60000

In [9]:
test_images.shape

(10000, 28, 28)

### The Network Architecture

In [12]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation = 'relu', input_shape = (28*28,)))
network.add(layers.Dense(10, activation = 'softmax')) # for 10 probabilty scores 


### Loss Function, Optimizer, Metrcis

The Compilation Step

In [14]:
network.compile(optimizer = 'rmsprop', 
                loss = 'categorical_crossentropy',
                metrics = ['accuracy'])


### Preparing the image data

In [15]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32')/255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32')/255

In [16]:
train_images

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

In [17]:
test_images

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

### Preparing the labels

In [19]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [21]:
train_labels

array([[0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.]], dtype=float32)

### Training the network

In [23]:
network.fit(train_images, train_labels, epochs = 3, batch_size = 128)

Epoch 1/3
469/469 [==============================] - 4s 9ms/step - loss: 0.1026 - accuracy: 0.9694
Epoch 2/3
469/469 [==============================] - 4s 9ms/step - loss: 0.0673 - accuracy: 0.9797
Epoch 3/3
469/469 [==============================] - 4s 9ms/step - loss: 0.0491 - accuracy: 0.9851


In [24]:
test_loss, test_acc = network.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 2ms/step - loss: 0.0626 - accuracy: 0.9802


### Insights on results

Here we see that the train set accuracy is about 98.51% and test set accuracy is about 98.02% Hence, problem of overfitting is not an issue here!